progress-indicator: Level 2 list item support (#1605)
agriculturegovau/agds-next: packages/react/src/progress-indicator/docs/overview.mdx

```jsx live
<ProgressIndicator
	activePath="/organisations"
	items={[

> 		{ label: 'Introduction', status: 'done', href: '/introduction' },

Reviewer: To prevent actually navigating away from the page, examples like SubNav put `#` at the start of the href url so it becomes a harmless hash link that still changes the path in the URL bar.

So for this it would look like 
`href: '#/introduction'`

Author: Hmm I think it's better to demonstrate how to use the component with folks' most common usage patterns. You can always click the back button to recover. Plus, if you click a different item and the state doesn't update it just seems broken. Plus plus, I'm wary of saying "it's totally fine to use in page links with this component".

What do you think about adding a separate example showing how to rig up `activePath` management, using the `#/introduction` technique you've mentioned and explicitly calling out not to do this in your app?

Reviewer 2: SideNav examples currently use hash links

Author: Looks like I'm in the minority here. Just confirming, we'll go to hash links as per @Dan503 's suggestion?

Reviewer 2: I would just do `#introduction` without the slash to match our other component examples. The slash feels weird

Author: - [ ] Move to hash links e.g. `#introduction`

```jsx live
<ProgressIndicator
	activePath="#organisations"
	items={[
		{ label: 'Introduction', status: 'done', href: '#introduction' },
		{ label: 'Submit evidence', status: 'saved', href: '#evidence' },
		{ label: 'Organisations', status: 'started', href: '#organisations' },
		{ label: 'Business contacts', status: 'error', href: '#contacts' },
		{ label: 'Case studies', status: 'todo', href: '#case-studies' },
		{ label: 'Attachments', status: 'started', href: '#attachments' },
		{ label: 'Review and submit', status: 'blocked', href: '#review' },
	]}
/>
```

> Progress indicators are used to help users navigate to each step of a multi-step form, show users how many steps are needed to complete and where they are in the process of completing it. This helps users know how much is left to do and how far along they are.

Reviewer 2: The second sentence feels unnecessary and repetitive of the first.

Also, I wonder if we should use "flow" instead of "form" since, not all pages are form pages (like Review and table file uploads etc)
```suggestion
Progress indicators are used to help users navigate to each step of a multi-step flow, show users how many steps are needed to complete and where they are in the process of completing it.
```

Author: @adrian-yee what are your thoughts on @stowball 's copy suggestions? (there are two suggestions)

Reviewer 3: we refer to the confirm and submit page and all secondary actions as part of a multi-step form - please revert to use 'Form':

Progress indicators are used to help users navigate to each step of a multi-step form, show users how many steps are needed to complete a from and how far they have progressed through the process of completing it.

Author: - [ ] Docs wording update

Progress indicators can be applied to larger tasks that can be completed in any order and returned to at a later date, like doing a tax return or applying for a driver’s licence.

Progress indicators can be applied to multi-step forms if the form has a known number of steps to be completed. <sup><a href="#cite-1">1</a></sup>

Avoid adding or removing pages to a multi-step form where a progress indicator is used as this can confuse and mislead users. Consider another approach if the number of steps might change due to user input. <sup><a href="#cite-2">2</a></sup>

> Progress indicators contain level 1 links to each page of a multi-step form. When additional pages are required for secondary actions nested in level 1 pages, level 2 links are revealed in the Progress indicator to help a user understand where they have navigated to and how to return to the to the previous step.

Reviewer 2: This feels confusing, verbose and has some typos. How about something like this:

```suggestion
By default, A Progress indicator contains links to each level 1 page of a multi-step flow.

When a user navigates to a secondary action page from a level 1 page, a nested level 2 link is revealed in the Progress indicator to help a user understand where they have navigated to and how to return to the previous step.
```

Author: Should the first sentence be `By default, a`? i.e. lowercase `a`?

Author: @adrian-yee what are your thoughts on @stowball 's copy suggestions?

		{ label: 'Organisations', status: 'started', href: '#organisations' },
		{ label: 'Business contacts', status: 'error', href: '#contacts' },
		{ label: 'Case studies', status: 'todo', href: '#case-studies' },
		{ label: 'Attachments', status: 'started', href: '#attachments' },
		{ label: 'Review and submit', status: 'blocked', href: '#review' },
	]}
/>
```

Progress indicators are used to help users navigate to each step of a multi-step form, show users how many steps are needed to complete a form and how far they have progressed through the process of completing it.

Progress indicators can be applied to larger tasks that can be completed in any order and returned to at a later date, like doing a tax return or applying for a driver’s licence.

Progress indicators can be applied to multi-step forms if the form has a known number of steps to be completed. <sup><a href="#cite-1">1</a></sup>

Avoid adding or removing pages to a multi-step form where a progress indicator is used as this can confuse and mislead users. Consider another approach if the number of steps might change due to user input. <sup><a href="#cite-2">2</a></sup>

> By default, a Progress indicator contains links to each level 1 page of a multi-step flow.

Reviewer 2: I suggested a change to "flow", but we use "form" everywhere else so we should change this back
```suggestion
By default, a Progress indicator contains links to each level 1 page of a multi-step form.
```

> By default, A Progress indicator contains links to each level 1 page of a multi-step flow.

Author: - [ ] lowercase `A`

	</TableBody>
</Table>

## Currently active item

To indicate the currently active item, whether level 1 or level 2, pass in a matching path to the `activePath` prop. E.g. `activePath="/organisations/change-name"`.

## Hiding the sub title

The Progress indicator component displays a subtitle "x of y steps completed" by default. To hide this subtitle, set the `hideSubtitle` prop to `true`.

```jsx live
<ProgressIndicator

> 	activePath="/organisations"

Reviewer: Can we make `activePath` optional?

If it is undefined, attempt to determine the `activePath` via the url. If it is explicitly defined then use that value to determine which href is active.

Author: This would make things inconsistent with our other `activePath` components.

Reviewer 2: I feel like this is another ticket to do this for all components with `activePath` if we decide it has value

	</TableBody>
</Table>

## Currently active item

> To indicate the currently active item, whether level 1 or level 2, pass in a matching path to the `activePath` prop. E.g. `activePath="/organisations/change-name"`.

Reviewer 2: ```suggestion
To indicate the currently active level 1 or level 2 item, pass in a matching path to the `activePath` prop, e.g. `activePath="/organisations/change-name"`.
```

## Hiding the sub title

The Progress indicator component displays a subtitle "x of y steps completed" by default. To hide this subtitle, set the `hideSubtitle` prop to `true`.

```jsx live
<ProgressIndicator
	activePath="#organisations"
	hideSubtitle
	items={[
		{ href: '#introduction', label: 'Introduction', status: 'done' },
		{ href: '#organisations', label: 'Organisations', status: 'started' },
		{
			href: '#contacts',
			label: 'Business contacts',
			status: 'started',
		},
		{ href: '#case-studies', label: 'Case studies', status: 'todo' },
		{ href: '#attachments', label: 'Attachments', status: 'blocked' },
	]}
/>
```

## Adding level 2 steps

> If one of the steps in a multi-step form has a secondary step, such as adding an employee with name and email to a list of employees, use a separate page for the secondary step. Let users know where they are by using a level 2 step in the progress indicator. A level 2 step sits under the main step. There can only be a single level 2 step for each main step.

Reviewer 2: This feels nicee to me:

```suggestion
If a level 1 page has a secondary step, such as editing its content, use a separate page for this secondary step.

When users navigate to this secondary step, let them know where they are by adding a nested level 2 step `item`. Only one level 2 step may exist for each level 1 step.
```

```jsx live
<ProgressIndicator
	activePath="#organisations/add-organisation"
	hideSubtitle
	items={[
		{ href: '#introduction', label: 'Introduction', status: 'done' },
		{
			href: '#organisations',
			label: 'Organisations',
			status: 'started',
			items: [
				{
					label: 'Add an organisation',
					href: '#organisations/add-organisation',
				},
			],
		},
		{ href: '#review', label: 'Review and submit', status: 'blocked' },
	]}
/>
```

> ### An example of a level 2 step
> 
> Progress indicators contain level 1 links to each page of a multi-step form. When additional pages are required for secondary actions nested in level 1 pages, level 2 links are revealed in the Progress indicator to help a user understand where they have navigated to and how to return to the to the previous step.
> 
> <ImageWithBorder src="/img/progress-indicator-level-two--1.png" alt="" />
> 
> The level 2 step is separate page, so it is displayed under the main level 1 step in the progress indicator. This helps the user understand where they are in the process.
> 
> <ImageWithBorder src="/img/progress-indicator-level-two--2.png" alt="" />
> 
> Once the user saves their changes, they’re taken back to the main step page and a section alert is displayed to let them know that their changes were made successfully. The level 2 step is no longer displayed in the progress indicator. It’s only displayed when the user is on the level 2 step page.
> 
> <ImageWithBorder src="/img/progress-indicator-level-two--3.png" alt="" />
> 

Reviewer 2: ```suggestion
```

I think you should remove the example since it contains lots of the same content and is not automated. That doesn’t have to be in this release

We can do that later as content updates get published immediately

Author: >I think you should remove the example since it contains lots of the same content and is not automated. That doesn’t have to be in this release
>
>We can do that later as content updates get published immediately

Yeah, I agree. The auto-generated images are a great concept because they stay up to date with code changes. Screenshots from local development or Figma don't. The next trade-off I see is either investing the time to create one off pages or using the `example-site` (which the [generate-template-image](https://github.com/agriculturegovau/agds-next/blob/main/docs/scripts/generateTemplateImages.js#L7) and [generate-guide-images](https://github.com/agriculturegovau/agds-next/blob/main/docs/scripts/generateGuideImages.js#L4) functions seems to use) or using the existing example site but running the risk of changes affecting the docs.

	hideSubtitle
	items={[
		{ href: '#introduction', label: 'Introduction', status: 'done' },
		{ href: '#organisations', label: 'Organisations', status: 'started' },
		{
			href: '#contacts',
			label: 'Business contacts',
			status: 'started',
		},
		{ href: '#case-studies', label: 'Case studies', status: 'todo' },
		{ href: '#attachments', label: 'Attachments', status: 'blocked' },
	]}
/>
```

## Adding level 2 steps

If a level 1 page has a secondary step, such as editing its content, use a separate page for this secondary step.

When users navigate to this secondary step, let them know where they are by adding a nested level 2 step `item`. Only one level 2 step may exist for each level 1 step.

```jsx live
<ProgressIndicator
	activePath="#organisations/add-organisation"
	hideSubtitle
	items={[
		{ href: '#introduction', label: 'Introduction', status: 'done' },
		{
			href: '#organisations',
			label: 'Organisations',
			status: 'started',
			items: [
				{
					label: 'Add an organisation',
					href: '#organisations/add-organisation',
				},
			],
		},
		{ href: '#review', label: 'Review and submit', status: 'blocked' },
	]}
/>
```

### An example of a level 2 step

> Progress indicators contain level 1 links to each page of a multi-step form. When additional pages are required for secondary actions nested in level 1 pages, level 2 links are revealed in the Progress indicator to help a user understand where they have navigated to and how to return to the to the previous step.

Reviewer 2: ```suggestion
By default, Progress indicators contain level 1 links to each page of a multi-step form. When additional pages are required for secondary actions nested under level 1 pages, level 2 links are revealed in the Progress indicator to help a user understand where they have navigated to and how to return to the previous step.
```

<ImageWithBorder src="/img/progress-indicator-level-two--1.png" alt="" />

The level 2 step is separate page, so it is displayed under the main level 1 step in the progress indicator. This helps the user understand where they are in the process.

<ImageWithBorder src="/img/progress-indicator-level-two--2.png" alt="" />

> Once the user saves their changes, they’re taken back to the main step page and a section alert is displayed to let them know that their changes were made successfully. The level 2 step is no longer displayed in the progress indicator. It’s only displayed when the user is on the level 2 step page.

Reviewer 2: We seem to always use a capital P elsewhere
```suggestion
Once the user saves their changes, they’re taken back to the main step page and a section alert is displayed to let them know that their changes were made successfully. The level 2 step is no longer displayed in the Progress indicator. It’s only displayed when the user is on the level 2 step page.
```

Author: Not everywhere 😅

/>
```

## Adding level 2 steps

> If one of the steps in a multi-step form has a secondary task, use a separate page for the secondary task. Let users know where they are by using a level 2 step in the progress indicator. A level 2 step sits under the main step. There can only be a single level 2 step for each main step.

Author: @Dan503 @stowball what do you think we should call these next level of steps? The options I'm seeing so far are:
1. Level 2 step - derived from the ticket + designs
2. Child step - as mentioned in the comment above
3. Sub step - I used this in a JSDoc accidentally but it falls in line with "sub nav" that we use

We should be consistent to help our consumers discover these features across different components, so I'm keen to get your thoughts!

> If one of the steps in a multi-step form has a secondary task, use a separate page for the secondary task. Let users know where they are by using a level 2 step in the progress indicator. A level 2 step sits under the main step. There can only be a single level 2 step for each main step.

Reviewer 2: I think sub step works best

```jsx live
<ProgressIndicator
	hideSubtitle
	items={[
		{ href: '#', label: 'Introduction', status: 'done' },
		{
			href: '#',
			label: 'Organisations',
			status: 'started',
			levelTwoItem: { label: 'Add an organisation', href: '#' },
		},
		{ href: '#', label: 'Review and submit', status: 'blocked' },
	]}
/>
```

### An example of a level 2 step

> Let’s say that a user is completing a multi-step application form. They want to change their pre-filled details, so they select the link to change their details.

Reviewer: You could use the YourGov employees list as an example. Saves you having to code up a scenario from scratch for the sake of screenshots.

Author: Nice idea! Could you please point me towards this list 🙏

Reviewer: Go here:
https://design-system.agriculture.gov.au/pr-preview/pr-1602/yourgov/

Sign in (any email and password works)

Charlies fast food

Apply for a permit

Fill out all the details in task 1

First page on Task 2 has the employee list
